File: week1/ResourceAllocation.java
/*
Question 5: Optimal Resource Allocation (Bitmask DP)

Given N tasks and N workers (N ≤ 20),
assign each worker exactly one task such that total cost is minimized.

Improve brute force O(N!) to O(N * 2^N) using state compression.
*/

import java.util.*;

class ResourceAllocation {

    public int findMinimumCost(int[][] costMatrix) {

        int totalTasks = costMatrix.length;
        int totalStates = 1 << totalTasks;

        int[] minimumCostForMask = new int[totalStates];
        Arrays.fill(minimumCostForMask, Integer.MAX_VALUE);

        minimumCostForMask[0] = 0;

        for (int taskMask = 0; taskMask < totalStates; taskMask++) {

            int workerIndex = Integer.bitCount(taskMask);

            for (int taskIndex = 0;
                 taskIndex < totalTasks;
                 taskIndex++) {

                if ((taskMask & (1 << taskIndex)) == 0) {

                    int newMask =
                            taskMask | (1 << taskIndex);

                    minimumCostForMask[newMask] =
                            Math.min(
                                    minimumCostForMask[newMask],
                                    minimumCostForMask[taskMask]
                                            + costMatrix[workerIndex][taskIndex]
                            );
                }
            }
        }

        return minimumCostForMask[totalStates - 1];
    }
}

File: week1/ResourceAllocation.cpp
/*
Question 5: Optimal Resource Allocation (Bitmask DP)

Given N tasks and N workers (N ≤ 20),
assign each worker exactly one task such that total cost is minimized.

Improve brute force O(N!) to O(N * 2^N) using state compression.
*/

#include <bits/stdc++.h>
using namespace std;

class ResourceAllocation {
public:
    int findMinimumCost(vector<vector<int>>& costMatrix) {
        int n = costMatrix.size();
        int totalStates = 1 << n;

        vector<int> dp(totalStates, INT_MAX);
        dp[0] = 0;

        for (int mask = 0; mask < totalStates; mask++) {

            int worker = __builtin_popcount(mask);

            for (int task = 0; task < n; task++) {

                if (!(mask & (1 << task))) {

                    int newMask = mask | (1 << task);

                    dp[newMask] = min(
                        dp[newMask],
                        dp[mask] + costMatrix[worker][task]
                    );
                }
            }
        }

        return dp[totalStates - 1];
    }
};
